Use of ESPAsyncTCP at 1.1.3 causes issues

<!-- Thanks for reporting a bug for this project. READ THIS FIRST:

- Provide as many details as possible. Simply saying "X gives bug" or "X gives error" is not enough!
- Paste logs, configuration sample and code into the backticks (```).
- Read through the template carefully and fill out all missing details.
- Please also search for similar issues in this issue tracker first and read through the ESPHome FAQ.

DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Operating environment/Installation (Hass.io/Docker/pip/etc.):**
macOS, no docker
<!--
Please provide details about your environment below the '-->'.
-->

**ESP (ESP32/ESP8266, Board/Sonoff):**
ESP8266 but makes no difference
<!--
Please provide details about which ESP you're using below.
-->

**Affected component:**
Generation of platformio.ini
<!--
Please add the link to the documentation at https://esphome.io/index.html of the component in question.
-->


**Description of problem:**
The ESPAsyncTCP is constrained to 1.1.3, this causes an install error:

```
Error: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/platformio/__main__.py", line 120, in main
    cli(None, None, None)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 700, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 680, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 873, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 508, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/platformio/commands/lib.py", line 103, in lib_install
    library, silent=silent, interactive=interactive, force=force)
  File "/usr/local/lib/python2.7/site-packages/platformio/managers/lib.py", line 352, in install
    force=force)
  File "/usr/local/lib/python2.7/site-packages/platformio/managers/package.py", line 723, in install
    pkg_dir = self._install_from_piorepo(name, requirements)
  File "/usr/local/lib/python2.7/site-packages/platformio/managers/lib.py", line 213, in _install_from_piorepo
    if app.get_setting("enable_ssl") else dl_data['url'], requirements)
  File "/usr/local/lib/python2.7/site-packages/platformio/managers/package.py", line 527, in _install_from_url
    return self._install_from_tmp_dir(_tmp_dir, requirements)
  File "/usr/local/lib/python2.7/site-packages/platformio/managers/package.py", line 601, in _install_from_tmp_dir
    pkg_dir = join(self.package_dir, target_dirname)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/posixpath.py", line 73, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 28: ordinal not in range(128)

============================================================

An unexpected error occurred. Further steps:

* Verify that you have the latest version of PlatformIO using
  `pip install -U platformio` command

* Try to find answer in FAQ Troubleshooting section
  https://docs.platformio.org/page/faq.html

* Report this problem to the developers
  https://github.com/platformio/platformio-core/issues

============================================================
```

And also some weird compile issues when TLS is enabeled 


```
ESPAsyncTCP_ID305/src/tcp_axtls.c:422:1: error: unknown type name 'bool'
```


Is there any particular reason why this constrain is used? when removing it and using version 1.2.0 it works just fine!




**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE

```

**Traceback (if applicable):**
<!--
Please copy the traceback here if compilation is failing. If possible, also connect to the ESP over USB and copy its logs into the backticks.
-->
```

```

**Additional information and things you've tried:**

@brandond 

Some additions here, when generating a yaml file with the wizard and running compile right after it the install of 305 works (interestingly)
When I then add mosquitto tls configuration I run into the strange compile errors. See the output here:

```
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiServerSecureAxTLS.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pioenvs/testinstall/libf99/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/AsyncPrinter.cpp.o
Compiling .pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/ESPAsyncTCP.cpp.o
Compiling .pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/ESPAsyncTCPbuffer.cpp.o
Compiling .pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/SyncClient.cpp.o
Compiling .pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/tcp_axtls.c.o
In file included from .piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.c:34:0:
.piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.h:87:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp);
^
.piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.c:422:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp){
^
*** [.pioenvs/testinstall/libb39/ESPAsyncTCP_ID305/tcp_axtls.c.o] Error 1
```

There is also a related closed issue to that on ESPAsyncTCP https://github.com/me-no-dev/ESPAsyncTCP/issues/82

I also have the issue where the compile fails when I try to use ssl_fingerprints with mqtt.
My setup is esphome installed using pip on a stock raspberry pi raspbian.
`Version: 1.12.1`

If you remove the ssl_fingerprints section below, then it all compiles fine.

Here is the yaml that causes the error and then the resulting error:

```
esphome:
  name: garage_temperature
  platform: ESP8266
  board: nodemcuv2

wifi:
  ssid: "SSID"
  password: "password"

# Enable logging
logger:

ota:
  password: 'password'

mqtt:
  broker: something.somewhere.com
  username: myusernmae
  password: mypassword
  port: 8883
  ssl_fingerprints:
    - FINGERPRINTHERE

sensor:
  - platform: dht
    pin: D2
    temperature:
      name: "Garage Temperature"
    humidity:
      name: "Garage Humidity"
    update_interval: 300s
```

Compiling causes:
```
|   |-- <AsyncMqttClient> 0.8.2
|   |   |-- <ESPAsyncTCP> 1.1.3
|   |   |   |-- <ESP8266WiFi> 1.0
Compiling .pioenvs/garage_temperature/src/main.cpp.o
Compiling .pioenvs/garage_temperature/lib764/ESPAsyncTCP_ID305/tcp_axtls.c.o
In file included from .piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.c:34:0:
.piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.h:87:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp);
^
.piolibdeps/ESPAsyncTCP_ID305/src/tcp_axtls.c:422:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp){
^
Compiling .pioenvs/garage_temperature/libc73/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PubRelPacket.cpp.o
*** [.pioenvs/garage_temperature/lib764/ESPAsyncTCP_ID305/tcp_axtls.c.o] Error 1
Compiling .pioenvs/garage_temperature/libc73/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PublishPacket.cpp.o
============================================= [ERROR] Took 10.55 seconds =============================================
```

I have the same problem and tried if the dev version (`sha256:b0d0f2f5a3d9abc716ff2b2078edfa6b1b35c84692ddbd4c028e4ca6c6e123f5`) of the esphome container would solve it, but it does not work. My yaml is mostly equivalent to @rwifall , using mqtt and the `ssl_fingerprints` option.

```
INFO Reading configuration...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d ringer
Processing ringer (platform: espressif8266@1.8.0; framework: arduino; board: d1_mini)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
LibraryManager: Installing id=346 @ 0.8.2
Downloading...
Unpacking  [####################################]  100%
LibraryManager: Installing id=305 @ ^1.1.0
Downloading...
Unpacking  [####################################]  100%
LibraryManager: Installing id=1826 @ ^1.0.0
Downloading...
Unpacking  [####################################]  100%
LibraryManager: Installing id=305 @ 1.1.3
Downloading...
Unpacking  [####################################]  100%
LibraryManager: Installing esphome-core
git version 2.17.1
Cloning into '/config/ringer/.piolibdeps/_tmp_installing-rTpZgu-package'...
remote: Enumerating objects: 400, done.
remote: Counting objects: 100% (400/400), done.
remote: Compressing objects: 100% (349/349), done.
remote: Total 400 (delta 51), reused 135 (delta 48), pack-reused 0
Receiving objects: 100% (400/400), 439.79 KiB | 1.74 MiB/s, done.
Resolving deltas: 100% (51/51), done.
LibraryManager: Installing id=3837 @ 5.13.3
Downloading...
Unpacking  [####################################]  100%
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/d1_mini.html
PLATFORM: Espressif 8266 > WeMos D1 R2 & mini
HARDWARE: ESP8266 80MHz 80KB RAM (4MB Flash)
Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF MODES: FINDER(chain) COMPATIBILITY(soft)
Collected 29 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <AsyncMqttClient> 0.8.2
|   |-- <ESPAsyncTCP> 1.2.0
|   |   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS>
|   |-- <ESP8266WiFi> 1.0
|-- <ESPAsyncTCP> 1.1.3
|   |-- <ESP8266WiFi> 1.0
|-- <Hash> 1.0
|-- <esphome-core> 1.13.0-dev #30da854
|   |-- <ArduinoJson-esphomelib> 5.13.3
|   |-- <ESP8266WiFi> 1.0
|   |-- <ESP8266mDNS>
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <AsyncMqttClient> 0.8.2
|   |   |-- <ESPAsyncTCP> 1.2.0
|   |   |   |-- <ESP8266WiFi> 1.0
|   |-- <ESPAsyncTCP> 1.2.0
|   |   |-- <ESP8266WiFi> 1.0
Compiling .pioenvs/ringer/src/main.cpp.o
Generating LD script .pioenvs/ringer/ld/eagle.app.v6.common.ld
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiServerSecureAxTLS.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pioenvs/ringer/lib4d9/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/ringer/libd45/ESPAsyncTCP_ID305/AsyncPrinter.cpp.o
Compiling .pioenvs/ringer/libd45/ESPAsyncTCP_ID305/ESPAsyncTCP.cpp.o
Compiling .pioenvs/ringer/libd45/ESPAsyncTCP_ID305/ESPAsyncTCPbuffer.cpp.o
Compiling .pioenvs/ringer/libd45/ESPAsyncTCP_ID305/SyncClient.cpp.o
Compiling .pioenvs/ringer/libd45/ESPAsyncTCP_ID305/tcp_axtls.c.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/ConnAckPacket.cpp.o
Archiving .pioenvs/ringer/lib4d9/libESP8266WiFi.a
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PingRespPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PubAckPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PubCompPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PubRecPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PubRelPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/PublishPacket.cpp.o
Archiving .pioenvs/ringer/libd45/libESPAsyncTCP_ID305.a
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/SubAckPacket.cpp.o
Compiling .pioenvs/ringer/lib20f/AsyncMqttClient_ID346/AsyncMqttClient/Packets/UnsubAckPacket.cpp.o
Compiling .pioenvs/ringer/lib0d3/ESP8266mDNS/ESP8266mDNS.cpp.o
Compiling .pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/AsyncPrinter.cpp.o
Compiling .pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/ESPAsyncTCP.cpp.o
Compiling .pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/ESPAsyncTCPbuffer.cpp.o
Compiling .pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/SyncClient.cpp.o
Compiling .pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/tcp_axtls.c.o
Compiling .pioenvs/ringer/lib6f5/Hash/Hash.cpp.o
Archiving .pioenvs/ringer/lib20f/libAsyncMqttClient_ID346.a
Compiling .pioenvs/ringer/lib6f5/Hash/sha1/sha1.c.o
In file included from .piolibdeps/ESPAsyncTCP_ID305@1.1.3/src/tcp_axtls.c:34:0:
.piolibdeps/ESPAsyncTCP_ID305@1.1.3/src/tcp_axtls.h:87:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp);
^
.piolibdeps/ESPAsyncTCP_ID305@1.1.3/src/tcp_axtls.c:422:1: error: unknown type name 'bool'
bool tcp_ssl_has(struct tcp_pcb *tcp){
^
*** [.pioenvs/ringer/lib1af/ESPAsyncTCP_ID305@1.1.3/tcp_axtls.c.o] Error 1
```

--- a/esphome/components/api.py
+++ b/esphome/components/api.py
@@ -86,7 +86,7 @@ def lib_deps(config):
     if CORE.is_esp32:
         return 'AsyncTCP@1.0.3'
     if CORE.is_esp8266:
-        return 'ESPAsyncTCP@1.1.3'
+        return 'ESPAsyncTCP@1.2.0'
     raise NotImplementedError
 
 

--- a/esphome/writer.py
+++ b/esphome/writer.py
@@ -288,7 +288,7 @@ def gather_lib_deps():
             lib_deps.add('AsyncTCP@1.0.1')
         lib_deps.add('ESPmDNS')
     elif CORE.is_esp8266:
-        lib_deps.add('ESPAsyncTCP@1.1.3')
+        lib_deps.add('ESPAsyncTCP@1.2.0')
         lib_deps.add('ESP8266mDNS')
 
     # avoid changing build flags order

